### ISSUE TITLE
plot_surface should shade at draw time after projection to axes space, instead of doing so in data space

### ISSUE THREAD

**Issue by anntzer** [matplotlib/matplotlib#18120]
### Bug report

**Bug summary**

```python
import matplotlib.pyplot as plt
import numpy as np

fig = plt.figure()
x, y = np.mgrid[-10:11, -10:11]
z = -(x ** 2 + y ** 2)

ax = fig.add_subplot(121, projection="3d")
pc = ax.plot_surface(x, y, z)
ax = fig.add_subplot(122, projection="3d")
pc = ax.plot_surface(x, y, z / 10)

plt.show()
```
gives
![test](https://user-images.githubusercontent.com/1322974/88836345-d1b92600-d1d6-11ea-8d41-818ae19c0abb.png)

In the right plot, the shading is basically constant throughout the image; in the left one, it changes very sharply at the tip of the paraboloid -- in fact, it nearly looks like the patches immediately at the left and the right are close to being the darkest and the brightest ones!  Given that these two patches have nearly the same orientation, how is that possible?

From a quick investigation, I believe the root of the issue is the following: the shading is computed as the dot product of the patch normal(\*) and the illumination direction, but this dot product is taken *in data space* (immediately when the Poly3DCollection is built in plot_surface) rather than after projection to "3d screen space" (at draw time).  In data space, the angle between the patches at the tip of the paraboloid is indeed much larger (the z-scale is much wider than the x/y-scale), hence the large change in dot product.

(\*) The patch normal is approximated using just three of the vertices of each patch, but this is not the source of the issue -- using e.g. the average of the normals computed at each vertex of the patch doesn't help.

Therefore, I believe the shading should instead be computed at draw time, after projection to "3d screen space".  As another argument in favor of such behavior, consider also that if we ever manage to support log-scale in mplot3d, we'd certainly have to compute the dot products after scale application (because they are even more meaningless before scale application).  As usual, this could possibly be opt-in via a flag... or perhaps not.

**Matplotlib version**

  * Operating system: linux
  * Matplotlib version: master
  * Matplotlib backend (`print(matplotlib.get_backend())`): qt5agg
  * Python version: 38
  * Jupyter version (if applicable):
  * Other libraries:

**Comment by jklymak** [comment]
I'm no shading expert, but usually there is the idea of a light source and a camera, and both have a location and elevation.  Usually you would want the elevation computed in data space, and therefore the steeper object would have sharper reflection?

**Comment by anntzer** [comment]
I tend to think that matplotlib "data space" is not physical space, physical space is instead computed from "data space" + scale + axes limits + axes box aspect ratio.
To reuse the log-scale example (again I realize we don't support that yet, but pretend we do), if doing the computation in data space, it would be possible to have two patches that are parallel in screen space but actually have different shading (because they are not parallel in data space).

**Comment by jklymak** [comment]
Right, and I'm saying the lighting models I have seen (i.e. Matlab) have been computed in data space so far as I can tell, not what you are calling physical space.  If I wanted to account for different aspect ratios, I'd move the elevation of the lighting source up or down.  

In this case, your light source hasn't changed location in data space, but your second plot is viewed from much lower in data space than your first.  I guess what you are saying is that you'd like the light source to follow the camera (which matlab calls a camlight).  https://www.mathworks.com/help/matlab/creating_plots/lighting-overview.html  Anyway, I don't think we should claim one way of placing the lights is "correct" and other ways a bug.  It seems that what you are asking for is a way to control the light behaviour, which seems like a really useful addition to mplot3d.

**Comment by scottshambaugh** [comment]
TODO: Revisit once https://github.com/matplotlib/matplotlib/pull/30980 which overhauls 3D transforms is merged
